diff --git a/src/http/modules/ngx_http_upstream_hash_module.c b/src/http/modules/ngx_http_upstream_hash_module.c
index f3e2b9d3..6e05100b 100644
--- a/src/http/modules/ngx_http_upstream_hash_module.c
+++ b/src/http/modules/ngx_http_upstream_hash_module.c
@@ -568,6 +568,15 @@ ngx_http_upstream_get_chash_peer(ngx_peer_connection_t *pc, void *data)
                 continue;
             }

+#if (NGX_HTTP_HEALTH_DETECT)
+            if (ngx_http_health_detect_upstream_check_peer_down(hp->rrp.peers->name, &peer->server, &peer->name)) {
+                continue;
+            }
+#endif
+
             if (peer->max_conns && peer->conns >= peer->max_conns) {
                 continue;
             }
diff --git a/src/stream/ngx_stream_upstream_hash_module.c b/src/stream/ngx_stream_upstream_hash_module.c
index 4fa9a2dc..103b4de1 100644
--- a/src/stream/ngx_stream_upstream_hash_module.c
+++ b/src/stream/ngx_stream_upstream_hash_module.c
@@ -9,6 +9,9 @@
 #include <ngx_core.h>
 #include <ngx_stream.h>

+#if (NGX_STREAM_HEALTH_DETECT)
+#include "ngx_stream_health_detect_module.h"
+#endif

 typedef struct {
     uint32_t                              hash;
@@ -237,6 +240,16 @@ ngx_stream_upstream_get_hash_peer(ngx_peer_connection_t *pc, void *data)
             goto next;
         }

+#if (NGX_STREAM_HEALTH_DETECT)
+        if (ngx_stream_health_detect_upstream_check_peer_down(hp->rrp.peers->name, &peer->server, &peer->name)) {
+            goto next;
+        }
+#endif
+
         if (peer->max_fails
             && peer->fails >= peer->max_fails
             && now - peer->checked <= peer->fail_timeout)
@@ -550,6 +563,17 @@ ngx_stream_upstream_get_chash_peer(ngx_peer_connection_t *pc, void *data)
                 continue;
             }

+#if (NGX_STREAM_HEALTH_DETECT)
+            if (ngx_stream_health_detect_upstream_check_peer_down(hp->rrp.peers->name, &peer->server, &peer->name)) {
+                continue;
+            }
+#endif
+
+
             if (peer->max_fails
                 && peer->fails >= peer->max_fails
                 && now - peer->checked <= peer->fail_timeout)
diff --git a/src/stream/ngx_stream_upstream_least_conn_module.c b/src/stream/ngx_stream_upstream_least_conn_module.c
index 739b20a9..48e72671 100644
--- a/src/stream/ngx_stream_upstream_least_conn_module.c
+++ b/src/stream/ngx_stream_upstream_least_conn_module.c
@@ -9,6 +9,9 @@
 #include <ngx_core.h>
 #include <ngx_stream.h>

+#if (NGX_STREAM_HEALTH_DETECT)
+#include "ngx_stream_health_detect_module.h"
+#endif

 static ngx_int_t ngx_stream_upstream_init_least_conn_peer(
     ngx_stream_session_t *s, ngx_stream_upstream_srv_conf_t *us);
@@ -143,6 +146,16 @@ ngx_stream_upstream_get_least_conn_peer(ngx_peer_connection_t *pc, void *data)
             continue;
         }

+#if (NGX_STREAM_HEALTH_DETECT)
+        if (ngx_stream_health_detect_upstream_check_peer_down(rrp->peers->name, &peer->server, &peer->name)) {
+            continue;
+        }
+#endif
+
         if (peer->max_fails
             && peer->fails >= peer->max_fails
             && now - peer->checked <= peer->fail_timeout)
@@ -198,6 +211,16 @@ ngx_stream_upstream_get_least_conn_peer(ngx_peer_connection_t *pc, void *data)
                 continue;
             }

+#if (NGX_STREAM_HEALTH_DETECT)
+        if (ngx_stream_health_detect_upstream_check_peer_down(rrp->peers->name, &peer->server, &peer->name)) {
+            continue;
+        }
+#endif
+
             if (peer->conns * best->weight != best->conns * peer->weight) {
                 continue;
             }
diff --git a/src/stream/ngx_stream_upstream_round_robin.c b/src/stream/ngx_stream_upstream_round_robin.c
index 36e2ec5c..60985f73 100644
--- a/src/stream/ngx_stream_upstream_round_robin.c
+++ b/src/stream/ngx_stream_upstream_round_robin.c
@@ -9,6 +9,10 @@
 #include <ngx_core.h>
 #include <ngx_stream.h>

+#if (NGX_STREAM_HEALTH_DETECT)
+#include "ngx_stream_health_detect_module.h"
+#endif
+

 #define ngx_stream_upstream_tries(p) ((p)->number                             \
                                       + ((p)->next ? (p)->next->number : 0))
@@ -104,6 +108,16 @@ ngx_stream_upstream_init_round_robin(ngx_conf_t *cf,
                 peer[n].down = server[i].down;
                 peer[n].server = server[i].name;

+#if (NGX_STREAM_HEALTH_DETECT)
+                if (!server[i].down) {
+                    ngx_stream_health_detect_upstream_add_peer(us, &server[i].name, &server[i].addrs[j]);
+                }
+#endif
+
                 *peerp = &peer[n];
                 peerp = &peer[n].next;
                 n++;
@@ -168,6 +182,16 @@ ngx_stream_upstream_init_round_robin(ngx_conf_t *cf,
                 peer[n].down = server[i].down;
                 peer[n].server = server[i].name;

+#if (NGX_STREAM_HEALTH_DETECT)
+                if (!server[i].down) {
+                    ngx_stream_health_detect_upstream_add_peer(us, &server[i].name, &server[i].addrs[j]);
+                }
+#endif
+
                 *peerp = &peer[n];
                 peerp = &peer[n].next;
                 n++;
@@ -452,6 +491,12 @@ ngx_stream_upstream_get_round_robin_peer(ngx_peer_connection_t *pc, void *data)
             goto failed;
         }

+#if (NGX_STREAM_HEALTH_DETECT)
+        if (ngx_stream_health_detect_upstream_check_peer_down(peers->name, &peer->server, &peer->name)) {
+            goto failed;
+        }
+#endif
+
         rrp->current = peer;

     } else {
@@ -557,6 +602,12 @@ ngx_stream_upstream_get_peer(ngx_stream_upstream_rr_peer_data_t *rrp)
             continue;
         }

+#if (NGX_STREAM_HEALTH_DETECT)
+        if (ngx_stream_health_detect_upstream_check_peer_down(rrp->peers->name, &peer->server, &peer->name)) {
+            continue;
+        }
+#endif
+
         peer->current_weight += peer->effective_weight;
         total += peer->effective_weight;